server/hwapi/data_models/repository.py

```diff
@@ -107,3 +111,253 @@ def get_or_create(
     db.commit()
     return instance, True
+
+
+def get_machines_with_same_hardware_params(
+    db: Session,
+    arch: str,
+    board: models.Device | None,
+    release: models.Release | None,
+) -> list[models.Machine]:
+    """
+    Retrieve all the machines that have the given architecture, motherboard (optionally),
+    and are certified for the given release (if specified)
+    """
+    return (
+        db.query(models.Machine)
+        .join(models.Certificate)
+        .join(models.Report, models.Certificate.reports)
+        .join(
+            models.device_report_association,
+            models.Report.id == models.device_report_association.c.report_id,
+        )
+        .join(
+            models.Device,
+            models.device_report_association.c.device_id == models.Device.id,
+        )
+        .filter(
+            and_(
+                models.Device.id == board.id if board else true,
+                models.Report.architecture == arch,
+                (
+                    models.Certificate.release_id == release.id
+                    if release is not None
+                    else true
+                ),
+            )
+        )
+        .distinct()
+        .all()
+    )
+
+
+def get_release_from_os(db: Session, os: OSValidator) -> models.Release:
+    """Return release object matching given OS data"""
+    release = (
+        db.query(models.Release)
+        .filter_by(release=os.version, codename=os.codename)
+        .first()
+    )
+    if release is None:
+        raise ValueError(
+            f"No matching release found for codename '{os.codename}', version '{os.version}'"
+        )
+    return release
+
+
+def clean_vendor_name(name):
+    """Remove "Inc"/"Inc." substring from vendor name and leading whitespaces"""
+    return name.replace("Inc.", "").replace("Inc", "").strip()
+
+
+def get_board_by_validator_data(
+    db: Session, board_validator: device_validators.BoardValidator
+) -> models.Device | None:
+    """Return device object (category==BOARD) matching given Board data"""
+    board = (
+        db.query(models.Device)
+        .join(models.Vendor)
+        .filter(
+            and_(
+                models.Vendor.name.in_(
+                    [
+                        board_validator.manufacturer,
+                        clean_vendor_name(board_validator.manufacturer),
+                    ]
+                ),
+                models.Device.name == board_validator.product_name,
+                models.Device.version == board_validator.version,
+                models.Device.category.in_(
+                    [DeviceCategory.BOARD.value, DeviceCategory.OTHER.value]
+                ),
+            )
+        )
+        .first()
+    )
+    return board
+
+
+def get_devices_by_machine_ids(db: Session, machine_ids: list[int]) -> Query:
+    """
+    Return query that contains Devices joined with Reports and Certificates for a
+    given list of machine IDs
+    """
```

Comment:
given that this function returns a query object. I'd rename it to something like `get_machines_devices_query`